src/Eloquent/Builder.php

```diff
@@ -207,8 +214,12 @@ public function createOrFirst(array $attributes = [], array $values = []): Model
             try {
                 $document = $collection->findOneAndUpdate(
                     $attributes,
-                    ['$setOnInsert' => $values],
-                    ['upsert' => true, 'new' => true, 'typeMap' => ['root' => 'array', 'document' => 'array']],
+                    ['$setOnInsert' => (object) $values],
```

Comment:
The object cast certainly doesn't hurt and is arguably correct in the event all field names were numeric (if that's even possible here).

I do think it's worth adding a comment to explain that you're intentionally keeping extra fields in `$values` (instead of filtering down to keys in the original param) to ensure compatibility with pre-5.0 server versions.